plugin-server/src/main/ingestion-queues/batch-processing/each-batch-ingestion.ts

```diff
@@ -150,20 +164,62 @@ export async function ingestEvent(
 let messageCounter = 0
 let messageLogDate = 0
 
-export function groupByTeamDistinctId(kafkaMessages: KafkaMessage[]): PipelineEvent[][] {
+function computeKey(pluginEvent: PipelineEvent): string {
+    return `${pluginEvent.team_id ?? pluginEvent.token}:${pluginEvent.distinct_id}`
+}
+
+async function emitToOverflow(queue: IngestionConsumer, kafkaMessages: KafkaMessage[]) {
+    await Promise.all(
+        kafkaMessages.map((message) =>
+            queue.pluginsServer.kafkaProducer.queueMessage(
+                {
+                    topic: KAFKA_EVENTS_PLUGIN_INGESTION_OVERFLOW,
+                    messages: [message],
+                },
+                true
+            )
+        )
+    )
+}
+
+export function splitIngestionBatch(
+    kafkaMessages: KafkaMessage[],
+    overflowMode: IngestionOverflowMode
+): [PipelineEvent[][], KafkaMessage[]] {
+    /**
+     * Prepares micro-batches for use by eachBatchParallelIngestion:
+     *   - events are parsed and grouped by token & distinct_id for sequential processing
+     *   - if overflowMode=Reroute, messages to send to overflow are in the second array
+     */
     const batches: Map<string, PipelineEvent[]> = new Map()
+    const toOverflow: KafkaMessage[] = []
     for (const message of kafkaMessages) {
+        if (overflowMode === IngestionOverflowMode.Reroute && message.key == null) {
+            // Overflow detected by capture, reroute to overflow topic
+            toOverflow.push(message)
+            continue
+        }
         const pluginEvent = formPipelineEvent(message)
-        const key = `${pluginEvent.team_id ?? pluginEvent.token}:${pluginEvent.distinct_id}`
-        const siblings = batches.get(key)
+        const eventKey = computeKey(pluginEvent)
+        if (overflowMode === IngestionOverflowMode.Reroute && !ConfiguredLimiter.consume(eventKey, 1)) {
+            // Local overflow detection triggering, reroute to overflow topic too
+            message.key = null
+            ingestionPartitionKeyOverflowed.labels(`${pluginEvent.team_id ?? pluginEvent.token}`).inc()
+            if (LoggingLimiter.consume(eventKey, 1)) {
+                status.warn('🪣', `Local overflow detection triggered on key ${eventKey}`)
+            }
+            toOverflow.push(message)
+            continue
+        }
+        const siblings = batches.get(eventKey)
         if (siblings) {
             siblings.push(pluginEvent)
         } else {
-            batches.set(key, [pluginEvent])
+            batches.set(eventKey, [pluginEvent])
         }
     }
 
-    return Array.from(batches.values())
+    return [Array.from(batches.values()), toOverflow]
```

Comment:
Can we return an object here rather than the tuple? I feel this gives a little more protection against potential mistakes, and gives a little bit of insight on the intend e.g. in the return typing.

Reply:
Will do!